mantidproject/mantid
# Indirect - QENS Fitting Interfaces - Selection of discontinuous spectra ranges

It is currently not possible to select discontinuous spectra ranges in the QENS Fitting Interfaces (ConvFit, IqtFit, MSDFit, JumpFit).

This PR adds a spectrum selection widget, designed in the MVP pattern, and implements its use in the QENS fitting interfaces. This widget allows for the selection of both continuous and discontinuous selection (through spinners and a line edit widget respectively). It also provides the option to mask the bins of individual spectrum; note that currently the mask should always be the same across spectra, since the only available fitting mode is sequential.

**To test:**
1. Navigate to Interfaces > Indirect > Data Analysis.
2. For the MSDFit, IqtFit, ConvFit and JumpFit interfaces, provide the supplied test files at the the top of the interface (for ConvFit, also supply the resolution file).
3. For MSDFit select a Gaussian model in the 'Fit Type' drop-down menu. Add 1 Exponential and a Stretched Exponential in the IqtFit interface. Select One Lorentzian in the ConvFit drop-down menu. Select 'Use Delta Function' in the ConvFit function browser.
4. For IqtFit, select '0' as the StartX and '0.12' as the EndX.
4. For each interface, select a continuous fitting range by ensuring 'Range' is selected in the 'Fit Spectra' drop-down menu and then using the two spinners to select any continuous range.
5. Click the 'Run' button for each of the fitting interfaces and ensure a fit is performed, for each of the selected spectra, and the results displayed for each fit.
6. For each interface, select a discontinuous fitting range by ensuring 'String' is selected in the 'Fit Spectra' drop-down menu and then enter a valid range string (first try an invalid string, e.g. out of bounds spectrum, and ensure an error message is displayed to the user).
7. Click the 'Run' button for each of the fitting interfaces and ensure a fit is performed, for each of the selected spectra, and the results displayed for each fit.
8. Perform a fit with either a continuous or discontinuous fitting range in each of the QENS fitting interfaces, this time adding a mask string (even number of comma-separated real numbers). This region will be excluded when a fit is performed.

**Test Files:**
[QENSFitTestFiles.zip](https://github.com/mantidproject/mantid/files/2116400/QENSFitTestFiles.zip)



Fixes #22064. 

Does this update require release notes?


---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
